source/common/runtime/runtime_impl.cc

```diff
@@ -555,9 +555,44 @@ void LoaderImpl::loadNewSnapshot() {
   tls_->set([ptr = std::move(ptr)](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return ptr;
   });
+  // Invalidate all locally tracked snapshots. These can not be deleted as the owning threads may
+  // be accessing them. Instead mark them invalid and clean up during the next in-thread call to
+  // snapshot()
+  {
+    absl::ReaderMutexLock lock(&snapshot_mutex_);
+    for (auto& it : snapshots_) {
+      it.second->valid_ = false;
+    }
+  }
 }
 
-const Snapshot& LoaderImpl::snapshot() { return tls_->getTyped<Snapshot>(); }
+const Snapshot& LoaderImpl::snapshot() {
+  if (tls_->currentThreadRegistered()) {
+    return tls_->getTyped<Snapshot>();
+  }
+
+  // Make sure that if a runtime snapshot is requested from outside of a worker thread, this does
+  // not crash Envoy. Instead, track thread to snapshot maps locally. In the common case, a
+  // snapshot will exist and will be valid, so snag a reader lock and check.
+  {
+    absl::ReaderMutexLock lock(&snapshot_mutex_);
+    auto snapshot_it = snapshots_.find(std::this_thread::get_id());
+    if (snapshot_it != snapshots_.end() && snapshot_it->second->valid_) {
+      return *snapshot_it->second->snapshot_;
+    }
+  }
+
+  // If this is the thread's first call to snapshot() or mergeFrom() has been called since the last
+  // call to snapshot() create a new snapshot.
+  {
+    absl::MutexLock lock(&snapshot_mutex_);
+    snapshots_.erase(std::this_thread::get_id());
+    auto snapshot_it =
+        snapshots_.emplace(std::this_thread::get_id(), new SnapshotState{createNewSnapshot(), true})
```

Comment:
I don't think this is thread safe as we expect `createNewSnapshot()` to run on the main thread. Here is how I would recommend implementing this:

- On the main thread, create/store a shared pointer to a snapshot that is to be used for non-registered threads. This can be done during the standard load call above, when you set valid to false in the map.
- In the code in this function, instead of creating a new snapshot, just grab the existing shared pointer and store it in the map under lock.
- Then I think the rest of the logic will remain the same.

WDYT? Edit: See above comment for a potentially simpler implementation.

Reply:
I'm not sure I understand the problem of createNewSnapshot being called on the main thread.  There's one lock, I don't think we deadlock, and as long as non-worker thread calls snapshot(), and latches it beyond the lifetime of the next call to snapshot() I think it's fine but maybe you can clue me in on slack

Reply:
The issue is `createNewSnapshot()` potentially reads from the disk, synchronizes with RTDS, etc. There can be concurrent access to all of these things, which is why today we always create the snapshot on the main thread and then fan it out using TLS. ]

IMO we can simply store the shared snapshot as something that is updated under write lock and read under read lock, and then offer a version of snapshot() that returns the shared pointer. This should be simple, will not change the current worker performance in any way, and will work fine for non-worker threads. If perf is an issue later for non-worker threads I think we can offer a better solution in which non-worker threads register for updates by providing some type of `Postable` interface?